crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs

```diff
@@ -0,0 +1,383 @@
+    }
+
+    let sender_idx = 3;
+    let should_recv_idx = 2;
+
+    let sender_location = &agents[sender_idx].2 .0[32..];
+
+    let mut kitsune_basis = KitsuneBasis::new(vec![0; 36]);
+    kitsune_basis.0[32..].copy_from_slice(&sender_location);
+    let basis = Arc::new(kitsune_basis);
+
+    // If the location was copied correctly then the basis location should be the same as the sender
+    // location. Due to the logic above, the receiver should have the sender's location in its arc.
+    assert_eq!(agents[sender_idx].2.get_loc(), basis.get_loc());
+
+    let test_op = TestHostOp::new(space.clone());
+
+    agents[sender_idx]
+        .0
+        .op_store()
+        .write()
+        .push(test_op.clone());
+
+    agents[sender_idx]
+        .1
+        .broadcast(
+            space.clone(),
+            basis.clone(),
+            KitsuneTimeout::from_millis(5_000),
+            BroadcastData::Publish {
+                source: agents[sender_idx].2.clone(),
+                op_hash_list: vec![test_op.into()],
+                context: FetchContext::default(),
+            },
+        )
+        .await
+        .unwrap();
+
+    tokio::time::timeout(std::time::Duration::from_secs(60), {
+        let op_store_recv = agents[should_recv_idx].0.op_store().clone();
+        async move {
+            loop {
+                if !op_store_recv.read().is_empty() {
+                    break;
+                }
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+        }
+    })
+        .await
+        .unwrap();
+
+    assert_eq!(1, agents[should_recv_idx].0.op_store().read().len());
+
+    for i in 0..5 {
+        if i == sender_idx || i == should_recv_idx {
+            continue;
+        }
+
+        // We've filtered out the sender and the receiver, who are expected to have the data.
+        // Now we check that the agent at the current index does not have the basis that the op was
+        // published to in its arc. That would make the test wrong, not Kitsune, so fail here!
+        let should_this_agent_hold_the_op = should_agent_hold_op_at_basis(&agents[i].0, agents[i].2.clone(), basis.clone());
+
+        // If this assertion fails, it means that the agent at index `i` has the basis in its arc which is not intended by the test setup.
+        assert!(!should_this_agent_hold_the_op, "Agent {i} should receive the data, this is a setup issue with the test");
+
+        // Now make the important assertion that the agent at index `i` did not receive the data! If it's not in the agents arc
+        // (which we just asserted above) then it should not have been received.
+        let store_lock = agents[i].0.op_store();
+        let store = store_lock.read();
+        assert!(
+            store.is_empty(),
+            "Agent {} should not have received any data but has {} ops",
+            i,
+            store.len()
+        );
+    }
```

Comment:
It's this bit I'd like to get a second pair of eyes on. I *think* that if the first check for 'basis in agent arc' passes then Kitsune being correct means that the op storage assertion shouldn't fail. However, it is flaky. That means that sometimes the op basis isn't in the agent's arc but it gets the op anyway.

This could mean that the test host implementation has bugs. It could mean that gossip and publish behave differently. It could mean that I've not configured Kitsune right and arcs are changing. It could just be that I've written code in this loop that doesn't make sense. 

Working out what is wrong is probably going to be difficult, so I'd like to at least check that my assumptions are reasonable!

Reply:
My sense at the moment is that this could be pointing to a behavior in holochain that wasn't explicitly intended. However, we wrote a lot of code assuming we'd someday have garbage collection for things when our arcs shrink, for example, so haven't been too careful about getting ops when we shouldn't.

Perhaps this falls into a class of optimizations that we should wait until after launch to address?